src/v/kafka/protocol/legacy_message.h

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "bytes/iobuf.h"
+#include "hashing/crc32.h"
+#include "kafka/server/logger.h"
+#include "model/record.h"
+#include "vlog.h"
+
+namespace kafka {
+
+/*
+ * Legacy messages are stored in message sets (effectively a packed array of
+ * serialized messags). There are two message types identified by magic values.
+ *
+ * Message V0
+ *   - offset
+ *   - length
+ *   - crc
+ *   - magic
+ *   - attributes
+ *
+ * Message V1
+ *   - Message V0
+ *   - timestamp
+ *
+ * Immediately after either message type is the optional key and optional value
+ * encoded with fixed width size prefix (standard kafka encoding). This differs
+ * from v2 which uses varint encoding.
+ *
+ * The crc covers the message bytes from the magic through the value. Unlike v2,
+ * legacy messages use crc32 not crc32c.
+ */
+struct legacy_message {
+    static constexpr uint16_t compression_mask = 0x07;
+
+    model::offset base_offset;
+    int32_t batch_length;
+    int32_t batch_crc;
+    int8_t magic;
+    int8_t attributes;
+    std::optional<model::timestamp> timestamp;
+    std::optional<iobuf> key;
+    std::optional<iobuf> value;
+
+    model::compression compression() const {
+        auto value = static_cast<uint8_t>(attributes) & compression_mask;
+        switch (value) {
+        case 0:
+            return model::compression::none;
+        case 1:
+            return model::compression::gzip;
+        case 2:
+            return model::compression::snappy;
+        case 3:
+            return model::compression::lz4;
+        default:
```

Comment:
Should we move the default case out of the switch, so that new unhandled compression types become a compile-time error? Although I guess there aren't going to be new compression types here.

Reply:
ah, yeh there won't be any new compression algos for these legacy formats. but ill fix this case statement. your suggestion is a better pattern!

Reply:
notice that the cases are integers derived from a bit mask. I think the compiler doesn't know much like it does with an enum class. I think it would say you didn't cover each of the 256 cases?